### PR TITLE
[Runtime] Lazily copy ivar list in initObjCClass.

### PR DESCRIPTION
**Pull request by mikeash**
There's often nothing that needs to be fixed up in the ivar list. In that case, we can avoid copying it. This saves time and memory, and allows the class rodata to be in immutable memory.

rdar://116189946

### CODE REVIEW

**Comment by mikeash** [comment]
@swift-ci please test

**Comment by mikeash** [comment]
@swift-ci please test linux platform

**Comment by aschwaighofer** [comment]
Could there be old binaries where the compiler did not follow the proper rules (make the fields writable if they are not const)?

In such a case would it be better to first copy and then set the field data to the new value?

**Comment by aschwaighofer** [comment]
I guess it would have been weird for that theoretical compiler to make the top-level RO struct non-const but the fields vector const.

**Comment by mikeash** [comment]
We still copy the list before we modify it, so that should be fine. The only actual difference here is that if we never modify the list, we don't copy it. There would be a potential issue there if something tries to modify the list later on, but nothing should ever do that.

**Comment by mikeash** [comment]
@swift-ci please test linux platform

File: stdlib/public/runtime/Metadata.cpp [hunk 1]
```diff
@@ -3441,38 +3441,45 @@ static void initObjCClass(ClassMetadata *self,
...
-      // computed, overwrite it and give it better type information.
-      if (ivar.Size != eltLayout->size) {
-        ivar.Size = eltLayout->size;
-        ivar.Type = nullptr;
-        ivar.Log2Alignment =
-          getLog2AlignmentFromMask(eltLayout->flags.getAlignmentMask());
+    ClassIvarEntry &ivar = ivars->getIvars()[i];
+
+    // Fill in the field offset global, if this ivar has one.
+    if (ivar.Offset) {
```
Comment (aschwaighofer) [inline stdlib/public/runtime/Metadata.cpp:3461]: Don't you modify it here? Before you copy it further down.

Comment (mikeash) [reply stdlib/public/runtime/Metadata.cpp:3461]: Not quite. We modify `*ivar.Offset` which is outside the list, and isn't getting copied anyway.

Comment (aschwaighofer) [reply stdlib/public/runtime/Metadata.cpp:3461]: Got it, thanks

**Review by aschwaighofer — APPROVED** [review]
LGTM

**Pull request merged**